paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FlinkRecordsWithSplitIds.java

```diff
@@ -113,8 +115,22 @@ public static void emitRecord(
             timestamp = metrics.getLatestFileCreationTime();
         }
 
+        // This metric only counts the number of RecordIterator<RowData> emitted,
+        // however what we really want is to count the number of RowData emitted,
+        // so we replenish the missing record count here.
+        org.apache.flink.metrics.Counter numRecordsIn =
+                context.metricGroup().getIOMetricGroup().getNumRecordsInCounter();
+        boolean firstRecord = true;
+
         RecordAndPosition<RowData> record;
         while ((record = element.next()) != null) {
+            // First record in the iterator is already counted by SourceReaderBase#pollNext.
+            if (firstRecord) {
+                firstRecord = false;
+            } else {
+                numRecordsIn.inc();
```

Comment:
Can we just count here, and invoke `Counter.inc(cnt - 1)` if cnt big than 1.

Reply:
It is possible that when emitting a batch of records, the source is backpressured and we have to stay in this while loop for some time. If at this moment the metric is queried, `numRecordsIn` will be inconsistent with that of the following operator.

I understand that we must be very careful when adding some code executed with each record. However `numRecordsIn.inc()` is quite simple. It just increases a `long` variable by 1.